packages/vitest/src/node/reporters/json.ts

```diff
@@ -60,20 +84,74 @@ export class JsonReporter implements Reporter {
 
     const success = numFailedTestSuites === 0 && numFailedTests === 0
 
-    const testResults: Array<TestResult> = tests.map(t => ({
-      perfStats: {
-        runtime: t.result?.duration,
-        start: t.result?.startTime,
-        end: t.result?.duration && t.result?.startTime && t.result.duration + t.result.startTime,
-      },
-      displayName: t.name,
-      failureMessage: t.result?.error?.message,
-      skipped: t.mode === 'skip',
-      status: t.result?.state,
-      testFilePath: t.file?.filepath,
-    }))
-
-    const result: AggregatedResult = { numTotalTestSuites, numPassedTestSuites, numFailedTestSuites, numPendingTestSuites, numTotalTests, numPassedTests, numFailedTests, numPendingTests, numTodoTests, startTime: this.start, success, testResults }
+    const testResults: Array<FormattedTestResult> = []
+    const fileToTestCases = new Map<string, { file: File; tests: Test[] }>()
+
+    for (const test of tests) {
+      const file = test.file
+      if (file) {
+        if (!fileToTestCases.has(file.filepath))
+          fileToTestCases.set(file.filepath, { file, tests: [] })
+
+        fileToTestCases.get(file.filepath)!.tests.push(test)
+      }
+    }
+
+    for (const [filepath, { tests, file }] of fileToTestCases) {
```

Comment:
I think that we could avoid calling `getTests` and then reconstruct the file-to-tests map if we iterate by file here. See getTests that will give you all the tests flatten for a file
https://github.com/vitest-dev/vitest/blob/a1922a07c54bf3f3b2cd9e2f23376ba254f97bab/packages/vitest/src/utils/tasks.ts#L4

```js
for (const file of files) {
  const tests = getTests(file)
```

Reply:
Yeah. You're right